File: javascript/test/property/graphs.test.js
const { Graph } = require('../../src/graphs/graph');
const { distances: bfsDistances } = require('../../src/graphs/bfs');
const { shortestPaths: dijkstra, INF } = require('../../src/graphs/dijkstra');
const { shortestPaths: bellmanFord } = require('../../src/graphs/bellmanFord');
const { allPairsShortestPaths } = require('../../src/graphs/floydWarshall');
const { totalWeight: kruskalWeight } = require('../../src/graphs/kruskalMst');
const { totalWeight: primWeight } = require('../../src/graphs/primMst');
const { totalWeight: boruvkaWeight } = require('../../src/graphs/boruvkaMst');
const { kahn } = require('../../src/graphs/topologicalSort');
const {
  PROPERTY_TRIALS,
  seededRng,
  bruteMstWeight,
} = require('../_propertyHelpers');

function randomConnected(rng, n, weighted) {
  const g = new Graph(n, false, weighted);
  for (let i = 1; i < n; i++) {
    const u = Math.floor(rng() * i);
    const w = weighted ? 1 + Math.floor(rng() * 9) : 1;
    g.addEdge(u, i, w);
  }
  for (let e = 0; e < n; e++) {
    const u = Math.floor(rng() * n);
    const v = Math.floor(rng() * n);
    if (u !== v) g.addEdge(u, v, weighted ? 1 + Math.floor(rng() * 9) : 1);
  }
  return g;
}

describe('property graphs', () => {
  test('BFS matches Dijkstra unweighted', () => {
    const rng = seededRng();
    for (let t = 0; t < PROPERTY_TRIALS; t++) {
      const n = 2 + Math.floor(rng() * 8);
      const g = randomConnected(rng, n, true);
      for (let u = 0; u < n; u++) {
        for (const e of g.neighbors(u)) e.weight = 1;
      }
      const src = Math.floor(rng() * n);
      const bfs = bfsDistances(g, src);
      const dij = dijkstra(g, src);
      for (let v = 0; v < n; v++) {
        if (bfs[v] === -1) expect(dij[v]).toBe(INF);
        else expect(dij[v]).toBe(bfs[v]);
      }
    }
  });

  test('Dijkstra matches Bellman-Ford', () => {
    const rng = seededRng();
    for (let t = 0; t < PROPERTY_TRIALS; t++) {
      const n = 2 + Math.floor(rng() * 6);
      const g = new Graph(n, true, true);
      for (let i = 1; i < n; i++) g.addEdge(i - 1, i, 1 + Math.floor(rng() * 5));
      const src = 0;
      const dij = dijkstra(g, src);
      const bf = bellmanFord(g, src);
      expect(bf.hasNegativeCycle).toBe(false);
      for (let v = 0; v < n; v++) expect(dij[v]).toBe(bf.distances[v]);
    }
  });

  test('Bellman-Ford negative cycle', () => {
    const rng = seededRng();
    for (let t = 0; t < PROPERTY_TRIALS; t++) {
      const g = new Graph(3, true, true);
      g.addEdge(0, 1, 1);
      g.addEdge(1, 2, -2);
      g.addEdge(2, 0, -2);
      expect(bellmanFord(g, 0).hasNegativeCycle).toBe(true);
    }
  });

  test('Floyd matches Dijkstra', () => {
    const rng = seededRng();
    for (let t = 0; t < PROPERTY_TRIALS; t++) {
      const n = 2 + Math.floor(rng() * 5);
      const g = randomConnected(rng, n, true);
      const fw = allPairsShortestPaths(g);
      for (let s = 0; s < n; s++) {
        const dij = dijkstra(g, s);
        for (let v = 0; v < n; v++) expect(fw[s][v]).toBe(dij[v]);
      }
    }
  });

  test('MST same weight', () => {
    const rng = seededRng();
    for (let t = 0; t < PROPERTY_TRIALS; t++) {
      const n = 3 + Math.floor(rng() * 6);
      const g = randomConnected(rng, n, true);
      const k = kruskalWeight(g);
      const p = primWeight(g);
      const b = boruvkaWeight(g);
      const brute = bruteMstWeight(n, g.allEdges());
      expect(k).toBe(p);
      expect(p).toBe(b);
      expect(b).toBe(brute);
    }
  });

  test('topological order', () => {
    const rng = seededRng();
    for (let t = 0; t < PROPERTY_TRIALS; t++) {
      const n = 2 + Math.floor(rng() * 8);
      const g = new Graph(n, true, false);
      for (let u = 0; u < n; u++) {
        for (let v = u + 1; v < n; v++) {
          if (rng() < 0.4) g.addEdge(u, v);
        }
      }
      const order = kahn(g);
      if (!order.length) continue;
      const pos = Object.fromEntries(order.map((v, i) => [v, i]));
      for (let u = 0; u < n; u++) {
        for (const e of g.neighbors(u)) {
          if (pos[u] !== undefined && pos[e.to] !== undefined) {
            expect(pos[u]).toBeLessThan(pos[e.to]);
          }
        }
      }
    }
  });
});

File: java/src/test/java/com/dsa/examples/graphs/GraphsTest.java
package com.dsa.examples.graphs;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class GraphsTest {

    @Test
    void bfsUnweighted() {
        Graph g = new Graph(5, false, false);
        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(1, 3);
        int[] dist = BFS.distances(g, 0);
        assertEquals(0, dist[0]);
        assertEquals(2, dist[3]);
        assertEquals(-1, dist[4]);
    }

    @Test
    void dfsAndTopo() {
        Graph dag = new Graph(4, true, false);
        dag.addEdge(0, 1);
        dag.addEdge(0, 2);
        dag.addEdge(1, 3);
        assertEquals(4, DFS.traverseRecursive(dag, 0).discovery().length);
        List<Integer> order = TopologicalSort.kahn(dag);
        assertEquals(4, order.size());
        assertEquals(0, order.get(0));
    }

    @Test
    void dijkstra() {
        Graph g = new Graph(4, true, true);
        g.addEdge(0, 1, 1);
        g.addEdge(1, 2, 2);
        g.addEdge(0, 2, 4);
        int[] dist = Dijkstra.shortestPaths(g, 0);
        assertEquals(3, dist[2]);
    }

    @Test
    void bellmanFordNegative() {
        Graph g = new Graph(3, true, true);
        g.addEdge(0, 1, 1);
        g.addEdge(1, 2, -2);
        g.addEdge(0, 2, 3);
        var result = BellmanFord.shortestPaths(g, 0);
        assertFalse(result.hasNegativeCycle());
        assertEquals(-1, result.distances()[2]);
    }

    @Test
    void floydWarshall() {
        Graph g = new Graph(3, true, true);
        g.addEdge(0, 1, 2);
        g.addEdge(1, 2, 3);
        int[][] dist = FloydWarshall.allPairsShortestPaths(g);
        assertEquals(5, dist[0][2]);
    }

    @Test
    void mstAlgorithms() {
        Graph g = new Graph(4, false, true);
        g.addEdge(0, 1, 1);
        g.addEdge(1, 2, 2);
        g.addEdge(2, 3, 3);
        g.addEdge(0, 3, 4);
        int kruskal = KruskalMST.totalWeight(g);
        int prim = PrimMST.totalWeight(g);
        int boruvka = BoruvkaMST.totalWeight(g);
        assertEquals(6, kruskal);
        assertEquals(6, prim);
        assertEquals(6, boruvka);
    }
}
